[clang][Parser] "Better" error messages for invalid template template

Veeloxfire opened the pull request:
For the somewhat easy mistake of `template<template A> ...` clang outputs a partially cryptic error

This patch changes this to assume the programmer intended `typename` in cases where `template` is illegal, and emit diagnostics (and forward parsing) accordingly

This mirrors the behaviour of `typedef` handling, so I feel its a reasonable qol change

Apologies if I missed any broken tests. Currently my local setup is messed up and fails on 100-200 tests eroniously, so I did my best to search through the errors and find the ones that needed updating

Veeloxfire committed:
[clang][Parser] "Better" error messages for invalid template template

For the somewhat easy mistake of `template<template A> ...` clang outputs a partially cryptic error
Change this to assume the programmer intended `typename` in cases where `template` is illegal, and emit diagnostics (and forward parsing) accordingly
This mirrors the behaviour of `typedef` handling

Endilll commented on clang/test/CXX/drs/cwg1xx.cpp:
```diff
@@ -1145,8 +1145,10 @@ namespace cwg181 { // cwg181: yes
   namespace X {
     template <template X<class T> > struct A { };
     // expected-error@-1 +{{}}
+    //  expected-note@-2 {{did you mean to use 'typename'?}}
```
```suggestion
    //   expected-note@-2 {{did you mean to use 'typename'?}}
```

Endilll commented on clang/test/CXX/drs/cwg1xx.cpp:
```diff
@@ -1145,8 +1145,10 @@ namespace cwg181 { // cwg181: yes
+    //  expected-note@-2 {{did you mean to use 'typename'?}}
     template <template X<class T> > void f(A<X>) { }
     // expected-error@-1 +{{}}
+    //  expected-note@-2 {{did you mean to use 'typename'?}}
```
```suggestion
    //   expected-note@-2 {{did you mean to use 'typename'?}}
```

Endilll reviewed (commented):
Changes to DR tests look fine.

Sirraide commented on clang/lib/Parse/ParseTemplate.cpp:
```diff
@@ -787,6 +787,23 @@ NamedDecl *Parser::ParseTemplateTemplateParameter(unsigned Depth,
                                                   unsigned Position) {
   assert(Tok.is(tok::kw_template) && "Expected 'template' keyword");
 
+  if (Token ahead = GetLookAheadToken(1);
+      ahead.isOneOf(tok::identifier, tok::ellipsis,
+                    tok::equal, tok::comma,
+                    tok::greater, tok::greatergreater, tok::greatergreatergreater)) {
```
I think it might be simpler to just check if the next token is *not* `<` and correct to `typename` in that case since a template-template parameter always has to start with `template <` anyway.

Veeloxfire replied:
In hindsight yes I think thats a better solution. I believe this solution came about because I was just going to check for `identifier` but then realised there were more options for type templates and kept adding

Sirraide reviewed (commented):
This does seem like a nice qoi improvement.

Veeloxfire committed:
[clang][Parser] Check for actual failure condition

`ahead.isNot(tok::less)` is the actual failure condition of the parse, so check for that instead of trying to match all the possible alternatives

Veeloxfire committed:
[clang][Parser] Reformat

Veeloxfire committed:
[clang][Parser] Reformat 2

Veeloxfire commented:
Sorry for the extra reformatting commit, I somehow missed the original formatting suggestions (I dont know how I didnt notice the 1 space indent originally)

Sirraide commented on clang/lib/Parse/ParseTemplate.cpp:
```diff
@@ -787,6 +787,20 @@ NamedDecl *Parser::ParseTemplateTemplateParameter(unsigned Depth,
                                                   unsigned Position) {
   assert(Tok.is(tok::kw_template) && "Expected 'template' keyword");
 
+  if (Token ahead = GetLookAheadToken(1); ahead.isNot(tok::less)) {
```
One more thing, I somehow forgot to comment on this earlier, but we typically use PascalCase for variables, so this should be `Ahead` (though I’d probably call it `Next` or sth, but that’s not that important...).

Veeloxfire replied:
Actually this prompted me to check and I was just writing `NextToken()` manually (didnt realise `GetLookAheadToken` did a `-1`), so I think I'll switch it to `NextToken()` instead.
Then the renaming makes sense too with `Token Next = NextToken();`

Sirraide replied:
Ah yes, you can tell I don’t review parser code too often because I probably should have noticed that haha.

Sirraide commented on clang/lib/Parse/ParseTemplate.cpp:
```diff
@@ -787,6 +787,20 @@ NamedDecl *Parser::ParseTemplateTemplateParameter(unsigned Depth,
   assert(Tok.is(tok::kw_template) && "Expected 'template' keyword");
 
+  if (Token ahead = GetLookAheadToken(1); ahead.isNot(tok::less)) {
+    // Maybe they intended `typename` instead of `template` (given thats more
+    // common) Error early, to add a fixit hint
```
```suggestion
    // `template` may have been a typo for `typename`, given that the
    // latter is more common.
```
nit: the comment could be cleaned up a bit, maybe like this

Veeloxfire replied:
Ah yes that comment is more consistent with other wording I've seen, thanks for the suggestion

Sirraide reviewed (approved):
LGTM after fixing two more minor things I only noticed just now.

Also, this needs a release note (in `clang/docs/ReleaseNotes.rst`, probably in the section ‘Improvements to Clang’s Diagnostics’ or whatever it was called).

Veeloxfire committed:
[clang][Parser] Suggested changes and cleanup

Veeloxfire commented:
> Also, this needs a release note (in `clang/docs/ReleaseNotes.rst`, probably in the section ‘Improvements to Clang’s Diagnostics’ or whatever it was called).

So I dont end up making 100 commits rewording this:
How about "Clang now diagnoses template template paramter missing `<` as typo for `typename`"

Sirraide commented:
> So I dont end up making 100 commits rewording this: 

It’s fine if you do. We do squash+merge anyway, so the entire pr will end up as one commit either way ;Þ

> How about "Clang now diagnoses template template paramter missing `<` as typo for `typename`"

Maybe something like ‘Clang now suggests a fix-it to correct `template` in a template-template parameter with a missing `<` to `typename`’ seeing as we were already diagnosing this; the only thing that’s changed is that we now also provide a fix-it.

Veeloxfire commented:
> It’s fine if you do. We do squash+merge anyway, so the entire pr will end up as one commit either way ;Þ

Aha I didnt realise!

> seeing as we were already diagnosing this; the only thing that’s changed is that we now also provide a fix-it.

I agree but also now it fully parses as if `typename` so that means code which previously did `template<template A> void foo(A)` will now parse `A` as a type (instead of skipping) and not error the second time for `A` not existing

This change isnt reflected in that, but I think its probably good enough and not worth worrying about

Sirraide commented:
> This change isnt reflected in that, but I think its probably good enough and not worth worrying about

Hmm, yeah, that’s true, but given that this is mainly about error correction, it’s not too important of a change imo; the program contains an error and won’t compile either way, after all.

Veeloxfire committed:
Updated release notes
